LIVY-356 Enabling LDAP authentication for Client to Server.

jakhani opened the pull request:
Currently, Livy doesn't support LDAP Authentication from client(sparkmagic) to server(livy). We need to add LDAP authentication as that's preferable method due to security reasons. We won't be able to use Knox for this purpose. That is why I am raising this PR which contains LDAP authentication.

Commit:
LIVY-356 Enabling LDAP authentication for Client to Server.

jakhani commented:
@jerryshao Can you please review this PR?

jerryshao commented:
Sure, I will.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+
+  private def hasDomain(userName: String) = indexOfDomainMatch(userName) > 0
+
+  private def indexOfDomainMatch(userName: String) = if (userName == null) -1
```
please change the style of this method to:

```
private def xxx(): Boolean = {
   
}
```

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  }
+}
+
+class LdapAuthenticationHandlerImpl() extends AuthenticationHandler {
```
Remove this unnecessary parenthesis `()`.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  private var enableStartTls: Boolean = false
+  private var disableHostNameVerification: Boolean = false
+
+  @VisibleForTesting def setEnableStartTls(enableStartTls: Boolean):
```
I don't see any use of this method here. Also there's no UT why do we need to add this annotation.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+
+object LdapAuthenticationHandlerImpl {
+  val AUTHORIZATION_SCHEME: String = "Basic"
+  val logger: Logger = LoggerFactory.getLogger(classOf[LdapAuthenticationHandlerImpl])
```
Don't create a logger here, instead using the `Logging` trait provided by Livy.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  def getType : String = "ldap"
+
+  @throws[ServletException]
+  def init(config: Properties): Unit = {
```
I'm a little confused for this method and below, looks like no one uses this method, not sure why implemented here.

jakhani replied:
init and destroy methods are inherited from AuthenticationHandler interface. These methods are getting called from AuthenticationFilter.

jerryshao reviewed (changes requested):
I remembered this functionality requires Hadoop version 2.8+, is that right?

Also please change the code, seems lots of code is not   used here.

jakhani commented on server/src/main/scala/org/apache/livy/LivyConf.scala:
```diff
@@ -96,6 +96,13 @@ object LivyConf {
+  // Ldap properties
+  val AUTH_LDAP_URL = Entry("livy.server.auth.ldap.url", null)
+  val AUTH_LDAP_BASE_DN = Entry("livy.server.auth.ldap.base-dn", null)
+  val AUTH_LDAP_USERNAME_DOMAIN = Entry("livy.server.auth.ldap.username-domain", null)
```
No. URL is must. I have added check for it in LdapAuthenticationHandlerImpl.scala.

Commit:
LIVY-356 Enabling LDAP authentication for Client to Server.

jakhani commented:
@jerryshao Earlier my plan was to use LDAP authentication supported by Hadoop auth library of 2.8+ version. But it didn't work well with our setup LDAP setup. Here I have implemented LDAP authentication which is working similar to that but with less restriction. I have fixed all the comments and also removed whatever code was not needed. There are couple of methods which are inherited from AuthenticationHandler interface. So I need to keep those. Init and Destroy methods are called from AuthenticationFilter. Can you please review recent changes?

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+
+object LdapAuthenticationHandlerImpl extends Logging {
+  val AUTHORIZATION_SCHEME = "Basic"
+  val TYPE = "org.apache.livy.server.auth.LdapAuthenticationHandlerImpl"
```
Is it better to be `val TYPE="ldap"`, otherwise 'livy.server.auth.type' should be configured with this long string?

jerryshao commented on server/src/main/scala/org/apache/livy/server/LivyServer.scala:
```diff
@@ -237,6 +238,24 @@ class LivyServer extends Logging {
         info(s"SPNEGO auth enabled (principal = $principal)")
 
+      case authType @ LdapAuthenticationHandlerImpl.TYPE =>
+        val holder = new FilterHolder(new AuthenticationFilter())
```
Should he be `new LdapAuthenticationHandlerImpl()`?

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+    if (userName == null) {
+      -1
+    }
+    else {
```
Can you please change all this if else clause to this style?

```
if () {
xxx
} else {
xxx
}

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  @throws[IOException]
+  @throws[AuthenticationException]
+  def managementOperation(token: AuthenticationToken, request: HttpServletRequest,
+                          response: HttpServletResponse) : Boolean = true
```
4 space indent.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  @throws[IOException]
+  @throws[AuthenticationException]
+  def authenticate(request: HttpServletRequest,
+                   response: HttpServletResponse): AuthenticationToken = {
```
4 space indent.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+      val credentials = new String(base64.decode(authorization),
+        StandardCharsets.UTF_8).split(":", 2)
+      if (credentials.length == 2) {
+        LdapAuthenticationHandlerImpl.debug(s"Authenticating  [${credentials(0)}] user")
```
Using Livy's logging trait.

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+
+import org.apache.livy._
+
+object LdapAuthenticationHandlerImpl extends Logging {
```
I don't see the usage of `Logging` in this object, why do you need to add here?

jerryshao commented on server/src/main/scala/org/apache/livy/server/auth/LdapAuthenticationHandlerImpl.scala:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @throws[AuthenticationException]
+  private def authenticateUser(userName: String, password: String) = if (
```
Can you please fix your style by following (https://github.com/databricks/scala-style-guide#parentheses)?

jerryshao reviewed (changes requested):
There're lots of style issue here, I'm afraid I cannot point all the issues one by one, can you please check Spark Scala style and fix all of them.

Besides I'm not familiar with Ldap, you need to make sure this code can be worked generally.

pranayhasan commented:
@jakhani Can you please address the issues raised?

ajbozarth commented:
@jakhani are you still willing to work on this? I know some users who would like to see this merged, but it still needs work.

jakhani commented:
Sorry for taking so much time in completing this PR. I am planning to work on this. I will submit review comment fixes in a week and close this PR.

ajbozarth commented:
Thanks @jakhani

AranVinkItility commented:
@jakhani It would really help me to see this great feature released. Can you give an update on what your plans are with this PR?

jerryshao commented:
The PR proposer seems doesn't work on this any more, and we don't have enough knowledge about LDAP, so this will be pending indefinitely.

pranayhasan commented:
@jerryshao @AranVinkItility Janki probably would address the comments by end of the month as she's out since 3 months. You can probably abandon this PR if it's not addressed by end of this month. Thanks for your patience.

lukatera commented:
any updates on this?

captainzmc commented:
Hi @jakhani, Now we need this part. If there are no updates, I will continue your work in the new PR and keep your previous commits.

jakhani commented:
Please go ahead. I won't be able to make this change. I have moved to
different team.

On Tue, Sep 10, 2019 at 12:22 AM micah zhao <notifications@github.com>
wrote:

> Hi @jakhani <https://github.com/jakhani>, Now we need this part. If there
> are no updates, I will continue your work in the new PR and keep your
> previous commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-livy/pull/20?email_source=notifications&email_token=AF5362IUHG4BDRFGDYOTFTDQI5DMHA5CNFSM4DUG776KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6KDT7Y#issuecomment-529807871>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF5362KNOCPTE3PYRESIMATQI5DMHANCNFSM4DUG776A>
> .
>


-- 
Thanks & Regards,
Janki Akhani

yuanzhaoYZ commented:
Thanks for adding this feature. But I couldn't find documentations on this topic. Posting my LDAP configs here. Hopefully it helps.

**Livy config**
```bash
# LDAP
livy.server.auth.type = ldap
livy.server.auth.ldap.url = ldap://localhost:389
livy.server.auth.ldap.base-dn = ou=people,dc=intellinum,dc=co
#livy.server.auth.ldap.username-domain = 
livy.server.auth.ldap.enable-start-tls = false
livy.server.auth.ldap.security-authentication = simple

```


**LDAP Config**
```bash
dn: uid=livy,ou=people,dc=intellinum,dc=co
objectclass: inetOrgPerson
cn: livy
sn: livy
uid: livy
userpassword: XXXXXX
ou: IT

````

wanggc2019 commented:
> Thanks for adding this feature. But I couldn't find documentations on this topic. Posting my LDAP configs here. Hopefully it helps.
> 
> **Livy config**
> 
> ```shell
> # LDAP
> livy.server.auth.type = ldap
> livy.server.auth.ldap.url = ldap://localhost:389
> livy.server.auth.ldap.base-dn = ou=people,dc=intellinum,dc=co
> #livy.server.auth.ldap.username-domain = 
> livy.server.auth.ldap.enable-start-tls = false
> livy.server.auth.ldap.security-authentication = simple
> ```
> 
> **LDAP Config**
> 
> ```shell
> dn: uid=livy,ou=people,dc=intellinum,dc=co
> objectclass: inetOrgPerson
> cn: livy
> sn: livy
> uid: livy
> userpassword: XXXXXX
> ou: IT
> ```

Thank you very much and may ask another question? My question is after enabling ldap, how to submit tasks with ldap authentication?
